DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/zero_copy_data_transfer/src/buffer_kernel.hpp

```diff
@@ -0,0 +1,55 @@
+#ifndef __BUFFER_KERNEL_HPP__
+#define __BUFFER_KERNEL_HPP__
+#pragma once
+
+#include <vector>
+
+#include <CL/sycl.hpp>
+#include <CL/sycl/INTEL/fpga_extensions.hpp>
+
+using namespace sycl;
+using namespace std::chrono;
+
+// Declare the kernel class name globally to avoid name mangling.
+class BufferWorker;
+
+// kernel using buffers to transfer data
+template <typename T>
+double BufferKernel(queue& q, std::vector<T>& in, std::vector<T>& out,
+                    const unsigned int size) {
+  // start timer
+  auto start = high_resolution_clock::now();
+
+  {
+    // set up the input/output buffers
+    buffer in_buf(in);
+    buffer out_buf(out);
+
+    // launch the computation kernel
+    auto kernel_event = q.submit([&](handler& h) {
+      accessor in_accessor { in_buf, h, read_only };
+      accessor out_accessor { out_buf, h, write_only, noinit };
+
+      h.single_task<BufferWorker>([=]() [[intel::kernel_args_restrict]] {
+        for (size_t i = 0; i < size; i++) {
+          out_accessor[i] = in_accessor[i] * i;
+        }
+      });
+    });
+  }
+
+  // We use the scope above to synchronize the FPGA kernels.
+  // Exiting the scope will cause the buffer destructors to be called
+  // which will wait until the kernels finish and copy the data back to the
+  // host (if the buffer was written to).
+  // Therefore, at this point in the code, we know the kernels have finished
+  // and the data has been transferred back to the host.
+
+  // stop the timer
+  auto end = high_resolution_clock::now();
+  duration<double, std::milli> diff = end - start;
```

Comment:
We should propose an extension of dpc_common.hpp timers such that you can ask for timer.Elapsed(std::milli);

This would allow use of the common headers.

Reply:
I agree. I proposed this originally, but I was ignored. I also proposed a helper function to grab sycl profiling info.

Reply:
You can make a pull request against the dpc_common.hpp file to propose those changes.